srcCpp03/PerftestTransport.cxx

```diff
@@ -265,14 +265,70 @@ bool configureWanTransport(
 
 void configureShmemTransport(
         PerftestTransport &transport,
+        dds::domain::qos::DomainParticipantQos &qos,
         std::map<std::string, std::string> &qos_properties,
         ParameterManager *_PM)
 {
-    int parentMsgSizeMax = atoi(
-            qos_properties["dds.transport.shmem.builtin.parent.message_size_max"]
-            .c_str());
+    using namespace rti::core::policy;
+    unsigned long long datalen = _PM->get<unsigned long long>("dataLen");
+    /*
+     * If we specify -scan, then we are interested in the highest size.
+     * Since the vector for scan is sorted, that number should be the last
+     * element.
+     */
+    if (_PM->is_set("scan")) {
+        const std::vector<unsigned long long> scanList =
+                _PM->get_vector<unsigned long long>("scan");
+        datalen = scanList[scanList.size() - 1];
+    }
+
+    long parentMsgSizeMax = transport.minimumMessageSizeMax;
+    std::ostringstream ss;
+    bool messageSizeMaxSet = false;
+
+    /*
+     * If the property defining the message_size_max for shared memory is not
+     * set and we are using exclusively SHMEM, then we will calculate
+     * automatically the message_size_max to accomodate the sample in a single
+     * packet and avoid fragmentation.
+     */
+    TransportBuiltinMask mask = qos.policy<TransportBuiltin>().mask();
+    if (qos_properties.find("dds.transport.shmem.builtin.parent.message_size_max")
```

Comment:
You can use
`if (qos_properties.count("dds.transport.shmem.builtin.parent.message_size_max")) {`

Reply:
cool, way better!